### DISCUSSION TITLE
Feature request: Context parameter for .query/.sql/.execute

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#10191] [Ideas]
duckdb's python bindings currently use the entire global and local namespace to search for dataframes. It would be nice to restrict this search via a `context: Dict[str, WhateverTypesDuckDBUnderstandsAsQueriables]` parameter.

(Btw, I've seen a  `conn.register` function but since I use duckdb entirely statelessly I'd like to just specify at each callsite what variables to use on that call)

EDIT: Not sure whether it'd make sense to consider merging the functionality of the `parameters`/`params` parameters in the sense of prepared statements and the variables that I'm describing here.

**Comment by soerenwolfers** [comment]
Once https://github.com/duckdb/duckdb/discussions/10962 is resolved, I can just write the required wrapper myself (unless I want a lazy query, which additionaly requires a resolution of https://github.com/duckdb/duckdb/issues/11687)

**Comment by Tishj** [comment]
> (Btw, I've seen a conn.register function but since I use duckdb entirely statelessly I'd like to just specify at each callsite what variables to use on that call)

Even then, you can call `duckdb.cursor` to create a one-off connection and `register` the desired variables there
That, combined with limiting the depth of stacks to scan to 0, would solve your issue

**Reply by soerenwolfers** [reply]
Ah, cool,  wasn't aware this is cursor specific / want aware what cursors are exactly

**Reply by Tishj** [reply]
They're part of DBAPI, I think in many implementations of the interface you *need* a cursor to do anything
Adding yet another step of boilerplate before you can do anything

In our implementation however they just create another connection, essentially equivalent* to:
```py
con1 = duckdb.connect(my_db_path)
con2 = duckdb.connect(my_db_path)
```

```py
con1 = duckdb.connect(my_db_path)
con2 = con1.cursor()
```

* they are equivalent for persistent databases, if using `:memory:` the DatabaseInstance that is created is not cached. Meaning another connection to `:memory:` will not use the same DatabaseInstance but instead create another.
Because of this we created `:memory:<name>` which *will* be cached.

All that to say: `cursor()` creates a new connection to the same underlying DatabaseInstance as the "parent" connection.

**Reply by soerenwolfers** [reply]
Ah, yeah, I was afraid there'd be some extra wasted effort in creating and dropping these on every call. Glad you thought about this
